internal/server/data/grant.go

```diff
@@ -106,8 +104,91 @@ func GetGrant(tx ReadTxn, opts GetGrantOptions) (*models.Grant, error) {
 	return (*models.Grant)(table), nil
 }
 
-func ListGrants(db GormTxn, p *Pagination, selectors ...SelectorFunc) ([]models.Grant, error) {
-	return list[models.Grant](db, p, selectors...)
+type ListGrantsOptions struct {
+	BySubject   uid.PolymorphicID
+	ByPrivilege string
+	ByResource  string
+
+	// IncludeInheritedFromGroups instructs ListGrants to include grants from
+	// groups where the user is a member. This option can only be used when
+	// BySubject is a non-zero userID.
+	IncludeInheritedFromGroups bool
+
+	// IncludeConnectorGrant instructs ListGrants to exclude grants where the
+	// CreatedBy is models.CreatedBySystem.
+	IncludeConnectorGrant bool
+
+	Pagination *Pagination
+}
+
+func ListGrants(tx ReadTxn, opts ListGrantsOptions) ([]models.Grant, error) {
+	table := grantsTable{}
+	query := querybuilder.New("SELECT")
+	query.B(columnsForSelect(table))
+	if opts.Pagination != nil {
+		query.B(", count(*) OVER()")
+	}
+	query.B("FROM grants")
+	query.B("WHERE deleted_at is null")
+	query.B("AND organization_id = ?", tx.OrganizationID())
+
+	if opts.BySubject != "" {
+		if !opts.IncludeInheritedFromGroups {
+			query.B("AND subject = ?", opts.BySubject)
+		} else {
+			subjects := []string{opts.BySubject.String()}
+
+			userID, err := opts.BySubject.ID()
+			if err != nil || !opts.BySubject.IsIdentity() {
+				return nil, fmt.Errorf("IncludeInheritedFromGroups requires a userId subject")
+			}
+			// FIXME: store userID and groupID as a field on the grants table so
+			// that we can replace this with a sub-select or join.
+			groupIDs, err := groupIDsForUser(tx, userID)
```

Comment:
This is the same implementation as we had before (a query for group membership, encoding the string ID in code, and then using those IDs in the main query). 

I think we're going to need to solve this problem for long-polling, so it should be a follow up.

Reply:
We discussed changing the grant's polymorphic fields from a single field into table + id fields; `subject` becomes `subject_type text` and `subject_id int`. Then you can rewrite this as a join through the membership table.